Kernel/ACPI/DMIDecoder.cpp

```diff
@@ -0,0 +1,257 @@
+            m_structures_count = m_entry64bit_point->table_maximum_size;
+            m_table_length = m_entry64bit_point->table_maximum_size;
+        } else if (m_entry32bit_point != nullptr) {
+            kprintf("DMIDecoder: SMBIOS 32bit Entry point @ P 0x%x\n", m_entry32bit_point.ptr());
+            m_use_64bit_entry = false;
+            m_structure_table = OwnPtr<SMBIOS::TableHeader>((SMBIOS::TableHeader*)m_entry32bit_point->legacy_structure.smbios_table_ptr);
+            m_structures_count = m_entry32bit_point->legacy_structure.smbios_tables_count;
+            m_table_length = m_entry32bit_point->legacy_structure.smboios_table_length;
+        }
+        kprintf("DMIDecoder: Data table @ P 0x%x\n", m_structure_table.ptr());
+        enumerate_smbios_tables();
+    } else {
+        m_operable = false;
+        kprintf("DMI Decoder is disabled. Cannot find SMBIOS tables.\n");
+    }
+}
+
+void DMIDecoder::enumerate_smbios_tables()
+{
+    u32 table_length = m_table_length;
+    SMBIOS::TableHeader* physical_table = m_structure_table.ptr();
+
+    auto region = MM.allocate_kernel_region(PAGE_ROUND_UP(table_length), "DMI Decoder Entry Point Finding", Region::Access::Read);
+    PhysicalAddress paddr = PhysicalAddress((u32)physical_table & PAGE_MASK);
+    mmap_region(*region, paddr);
+
+    auto* v_smbios_table = (SMBIOS::TableHeader*)(region->vaddr().get() + ((u32)physical_table & (~PAGE_MASK)));
+
+    u32 structures_count = 0;
+    while (table_length > 0) {
+#ifdef SMBIOS_DEBUG
+        dbgprintf("DMIDecoder: Examining table @ P 0x%x\n", physical_table);
+#endif
+        structures_count++;
+        if (v_smbios_table->type == (u32)SMBIOS::TableType::EndOfTable) {
+            kprintf("DMIDecoder: Detected table with type 127, End of SMBIOS data.\n");
+            break;
+        } else {
+            m_smbios_tables.append(physical_table);
+            table_length -= v_smbios_table->length;
+            kprintf("DMIDecoder: Detected table with type %u\n", v_smbios_table->type);
+        }
+        SMBIOS::TableHeader* physical_next_table = get_next_physical_table(*physical_table);
+#ifdef SMBIOS_DEBUG
+        dbgprintf("DMIDecoder: Next table @ P 0x%x\n", physical_next_table);
+#endif
+        if (physical_next_table == nullptr) {
+            break;
+        } else {
+            v_smbios_table = (SMBIOS::TableHeader*)(region->vaddr().get() + (u32)physical_next_table - paddr.get());
+            physical_table = physical_next_table;
+        }
+    }
+    m_structures_count = structures_count;
+}
+
+size_t DMIDecoder::get_table_size(SMBIOS::TableHeader& table)
+{
+    // FIXME: Make sure we have some mapping here so we don't rely on existing identity mapping...
+    size_t i;
+#ifdef SMBIOS_DEBUG
+    dbgprintf("DMIDecoder: table legnth - 0x%x\n", table.length);
+#endif
+    const char* strtab = (char*)&table + table.length;
+    for (i = 1; strtab[i - 1] != '\0' || strtab[i] != '\0'; i++)
```

Comment:
Should this loop not be bounded somehow? Is it guaranteed to always be null terminated?

Reply:
I agree it's not safe. The best option is to use the structure table size (you get this information from the entry point of SMBIOS). If the SMBIOS is really buggy, then we will need to use that info to make sure we can bail out before crashing.